solr/core/src/java/org/apache/solr/filestore/NodeFileStore.java

```diff
@@ -0,0 +1,211 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.filestore;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.apache.solr.handler.ReplicationHandler.FILE_STREAM;
+import static org.apache.solr.response.RawResponseWriter.CONTENT;
+import static org.apache.solr.security.PermissionNameProvider.Name.FILESTORE_READ_PERM;
+
+import jakarta.inject.Inject;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.solr.api.JerseyResource;
+import org.apache.solr.client.api.endpoint.NodeFileStoreApis;
+import org.apache.solr.client.api.model.FileStoreDirectoryListingResponse;
+import org.apache.solr.client.api.model.FileStoreEntryMetadata;
+import org.apache.solr.client.api.model.FileStoreJsonFileResponse;
+import org.apache.solr.client.api.model.SolrJerseyResponse;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.params.CommonParams;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.core.SolrCore;
+import org.apache.solr.jersey.PermissionName;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Implementation for {@link NodeFileStoreApis} */
+public class NodeFileStore extends JerseyResource implements NodeFileStoreApis {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private final CoreContainer coreContainer;
+  private final SolrQueryRequest req;
+  private final SolrQueryResponse rsp;
+  private final FileStore fileStore;
+
+  @Inject
+  public NodeFileStore(
+      CoreContainer coreContainer,
+      DistribFileStore fileStore,
+      SolrQueryRequest req,
+      SolrQueryResponse rsp) {
+    this.coreContainer = coreContainer;
+    this.req = req;
+    this.rsp = rsp;
+    this.fileStore = fileStore;
+  }
+
+  // TODO - this single "get" operation actually supports several different chunks of functionality:
```

Comment:
I wish we had JIRA's that referenced the "Todo" work...

Reply:
It would be nice.  I think the lack of a JIRA reference in this case reflects my own uncertainty about how to proceed.

In theory SOLR-17302 is already sortof meant to cover this type of "cosmetic improvement" change. That's how many of our v2 tickets are setup - to cover both JAX-RS conversion and "cosmetic changes" all in one.

But in practice I'm not sure whether it's safe/responsible to modify the appearance of these particular v2 endpoints.  They're technically "experimental", but they're the only way to access the filestore - there's no v1 equivalent that we can recommend to folks.

I'm going to start a discussion on SOLR-17302 about this, which will hopefully draw some more eyes and get us some clarity on how to proceed.

Reply:
Actually - small correction - apparently when I wrote up SOLR-17302 I worded it in such a way that it only covers migration to JAX-RS.

So there's no ticket (yet) for breaking up this API into more manageable chunks.  I'll file a separate ticket for this and reference it here (and in the code) when its ready.

Reply:
FYI: https://issues.apache.org/jira/browse/SOLR-17351